cci_cameroon/analysis/text_similarity/similarity_search_sample.py

```diff
@@ -0,0 +1,608 @@
+
+# %%
+index3.train(sentence_embeddings)
+index3.add(sentence_embeddings)
+
+# %%
+index3.nprobe = 1  # nprobe specifies the number of neigboring cells to be used
+
+# %%
+# we now perform a search for similarity using the input text read earlier on.
+# %time
+distance3, position3 = index3.search(xq, k)  # search
+# get the text
+for i in position3[0]:
+    print(sentences[i])
+print(distance3[0])
+
+# %%
+sentences[position[0][0]]
+
+# %%
+sentences[position[0][1]]
+
+# %%
+print("Similarity:", util.dot_score(xq, sentence_embeddings)[0][position[0][1]])
+
+# %% [markdown]
+# ## Output summary
+# For the given dataset, we notice that the algorithm returns the same set of indexes. However, when compression of the vectors happens, the indexes are returned in a reversed order. This shows that accuracy is compromised. The least distance between the vectors increases from 90 to 130.
+
+# %%
+# pip install hydra-core omegaconf
+
+# %% [markdown]
+# # Using the distilbert model for embedding
+
+# %%
+# using the distilbert model for embedding
+model2 = SentenceTransformer(
+    "distilbert-base-nli-stsb-mean-tokens"
+)  # best performance for all distillbert pretrained models.
+
+# %%
+# check GPU availability and set cuda if present.
+if torch.cuda.is_available():
+    model2 = model2.to(torch.device("cuda"))
+print(model2.device)
+
+# %%
+# produce sentence vectors
+embeddings = model2.encode(sentences)
+
+# %%
+# Step 1: Change data type
+embeddings2 = np.array([embedding for embedding in embeddings]).astype("float32")
+
+# %%
+print(embeddings[:10])
+
+# %%
+print(embeddings2[:10])
+
+# %%
+# create a faiss index for the search using the embenddings
+index_dis = faiss.IndexFlatL2(embeddings.shape[1])
+
+# %%
+index_dis.add(embeddings)  # add the vector to the index created
+
+# %%
+# search the index for our given input xq
+# %time
+distance4, position4 = index_dis.search(xq, k)  # search
+# get the text
+for i in position4[0]:
+    print(sentences[i])
+print(distance4[0])
+
+# %%
+sentences[position4[0][0]]
+
+# %%
+num_cells = 50  # Number of cells the index should be divided into
+quantizer = faiss.IndexFlatL2(dimension)
+index_dis2 = faiss.IndexIVFFlat(quantizer, dimension, num_cells)
+
+# %%
+index_dis2.train(embeddings)
+index_dis2.add(embeddings)
+
+# %%
+# search the index for our given input xq
+# %time
+distance5, position5 = index_dis2.search(xq, k)  # search
+# get the text
+for i in position5[0]:
+    print(sentences[i])
+print(distance5[0])
+
+# %%
+# print("Similarity:", util.dot_score(xq, embeddings)[0][5168])
+# We use cosine-similarity and torch.topk to find the highest 5 scores
+# cos_scores = util.pytorch_cos_sim(xq, embeddings)[0]
+# top_results = torch.topk(cos_scores, k=k)
+
+# %% [markdown]
+# # Using French models
+# We create a sample corpus of French sentences that would be used across the different pre-trained models below.
+
+# %%
+sentences_french = [
```

Comment:
I would also store this in inputs/data as a text file and call it in using a function stored in data getters. Maybe just add a short markdown description of what the text is.

Reply:
Or could actually be a dictionary / list of dictionaries stored as a json (key and value pair to group by sample test or corpus maybe)